### DISCUSSION TITLE
Why can't I use `onChange` from `register()` passing as a prop?

### DISCUSSION THREAD

**Discussion by hyungjunk** [react-hook-form/react-hook-form#8704] [Q&A]
I created an **input file type** component which has a ground validation rule that goes for all cases.

For example, if a file type is not accepting only png type but everything else, the component I created would be like

```
const InputFile = forwardRef<HTMLInputElement, InputFileProps>((props, ref) => {
  const { onChange } = props;
  return (
    <input
      ref={ref}
      type="file"
      hidden={true}
      {...props}
      onChange={(e: ChangeEvent<HTMLInputElement>) => {
        const eventWithFilteredFiles = filterOutPNGFiles(e);
        onChange?.(eventWithFilteredFiles);
      }}
    />
  );
});
```

This only works for components where a react-hook-form`register()` is not used.

 I think I could pass `onChange` method as a register option, but still wonder how to achieve and why it doesn't work the way I expected. Thanks

**Comment by bluebill1049** [comment]
mostly due to onChange over right. you can use `register('test', { onChange })` or combine register onChange and your custom onChange.

```
onChange={(e) => {
  props.onChange() // your own props
  onChange() // from register
}}

**Reply by Moshyfawn** [reply]
The last passed `onChange` overrides any previously passed `onChange`. That's the way React JSX works, nothing specific to RHF

Check our what props `register` returns when you spread it into your input

**Comment by hyungjunk** [comment]
@bluebill1049 @Moshyfawn Thanks for the answers. I think I missed the use case of the component in my question so I am attaching it here and how I use it. 

To simplify, it looks like this.
```
<InputFile
  {...register('myProfile')}
  id="myProfile"
  accept=""
/>
```
As you can see, `onChange` method is what came from register method. It is not overriden right?

**Reply by Moshyfawn** [reply]
Yep, if you don't pass another `onChange` to your input after spreading `register` props, it's not overriden

If you want to run some additional `onChange` logic, pass it into the second `register` `onChange` prop ;)

**Reply by hyungjunk** [reply]
Yeah, I understand, but I am trying to make a single component where `onChange` validation logic is defined only here so I can reuse this component whenever I need a file input component. So it turns out that it is not a matter of `onChange` overriding. What would make `onChange` method not work here?
